### PR TITLE
Fixing swagger spec

### PR DESCRIPTION
**Pull request by nikhiljindal**
Fixes https://github.com/swagger-api/swagger-codegen/issues/1066

Updating typeToJSON() to consider "[]string". Also changing type.Name() to type.String() since type.Name() returns an empty string for "[]string"

### CODE REVIEW

**Comment by nikhiljindal** [comment]
cc @bgrant0607 @brendandburns

**Comment by nikhiljindal** [comment]
cc @madhuri-rai07

File: api/swagger-spec/v1.json [hunk 1]
```diff
@@ -5971,7 +5971,7 @@
         "allowMultiple": false
        },
        {
-        "type": "",
+        "type": "string",
```
Comment (bgrant0607) [inline api/swagger-spec/v1.json]: This is supposed to be an array of strings. How does one represent that?

Comment (wing328) [inline api/swagger-spec/v1.json]: @bgrant0607 for swagger spec 1.2, you can use `allowMultiple` for query parameter to represent multiple values:

https://github.com/swagger-api/swagger-spec/blob/master/versions/1.2.md#524-parameter-object

Comment (bgrant0607) [inline api/swagger-spec/v1.json]: Thanks!

@nikhiljindal In the case of an array, we need to call AllowMultiple on the parameter:
https://github.com/emicklei/go-restful/blob/bdfb7d41639a84ea7c36df648e5865cd9fbf21e2/parameter.go#L81

Comment (wing328) [inline api/swagger-spec/v1.json]: Please also note that it assumes the values are comma-separated so the server should expect comma-separated values. 

Swagger Spec 2.0 has more options on the format. Please refer to https://github.com/swagger-api/swagger-spec/blob/master/versions/2.0.md#fixed-fields-7 => `collectionFormat` for more information

Comment (bgrant0607) [inline api/swagger-spec/v1.json]: There's too much automagic in queryToObject for me to easily determine whether we split by comma or just expect multiple args of the same name or what.

cc @ncdc

Comment (ncdc) [inline api/swagger-spec/v1.json]: I'm not super familiar with queryToObject but can take a look. What's the specific issue? For exec, we expect multiple args of the same name, or at least we did when I originally wrote exec.

cc @smarterclayton

Comment (nikhiljindal) [inline api/swagger-spec/v1.json]: An array of strings can be represented as:
```
"type": "array",
      "items": {
       "type": "string"
      },
```

**Comment by nikhiljindal** [comment]
Any ideas on how can I test passing multiple commands to exec on API server?
```kubectl exec``` supports only one command.

running curl pods/exec gives me a 404.
```
$ curl http://localhost:8080/api/v1/namespaces/default/pods/my-nginx-alrxm/exec
{
  "kind": "Status",
  "apiVersion": "v1",
  "metadata": {},
  "status": "Failure",
  "message": "Upgrade request required",
  "reason": "BadRequest",
  "code": 400
}
```

**Comment by ncdc** [comment]
It's just 1 command. The http request includes multiple values for the
command query param - to make it an array of args

**Comment by nikhiljindal** [comment]
aah yes you are right.

I ran
```
./cluster/kubectl.sh exec my-nginx-fs9uw echo "ABC" "def"
```
which sent the following request to APIserver:
```
 http://localhost:8080/api/v1/namespaces/default/pods/my-nginx-fs9uw/exec?command=echo&command=ABC&command=def&container=my-nginx&stderr=true&stdout=true
```

So we want multiple args of the same name

File: api/swagger-spec/v1.json (hunk 1 above)
Comment (nikhiljindal) [inline api/swagger-spec/v1.json]: @ncdc You are right. I verified that exec expects multiple args of the same name (multiple command query params, in this case) of the form: exec?command=echo&command=ABC

@wing328 Do you know whats the best way to represent that?

Comment (wing328) [inline api/swagger-spec/v1.json]: @nikhiljindal Swagger 1.2 spec does not support that (multiple args of the same name) based on my knowledge.

[Swagger 2.0](https://github.com/swagger-api/swagger-spec/blob/master/versions/2.0.md#fixed-fields-7) spec supports that with `collectionFormat` using `multi`:

```
multi - corresponds to multiple parameter instances instead of multiple values 
for a single instance foo=bar&foo=baz. This is valid only for parameters in 
"query" or "formData"
```

cc @webron - the expert in Swagger Spec

Comment (webron) [inline api/swagger-spec/v1.json]: Oh wow, that feels like bringing up something to life from the dead. That part of the spec wasn't well defined in 1.2 which is why it was revised in 2.0. I honestly don't recall if `multi` as it is in 2.0 existed in 1.2, but I believe it did not (was one of the feature requests).

Comment (bgrant0607) [inline api/swagger-spec/v1.json]: This is just going to be wrong, then, until we update to 2.0.

Comment (nikhiljindal) [inline api/swagger-spec/v1.json]: Or I can try updating exec to accept both
```
exec?command=echo&command=ABC
``` 
and
```
exec?command=echo,ABC
```
(the second case will not work for commands with ',' tough)

**Comment by nikhiljindal** [comment]
Have added another commit to convert csv query params to []strings before converting url.Values to struct. The conversion is done only when the destination struct field has a "csv" json tag.

@bgrant0607 PTAL and let me know what you think of it.
With that commit, both of  the following will work:

```
exec?command=echo&command=ABC
```
and
```
exec?command=echo,ABC
```

**Comment by nikhiljindal** [comment]
@k8s-bot ok to test
Please pass this time :)

File: pkg/api/v1/types.go [hunk 2]
```diff
@@ -1572,7 +1572,7 @@ type PodExecOptions struct {
 	Container string `json:"container,omitempty" description:"the container in which to execute the command. Defaults to only container if there is only one container in the pod."`
 
 	// Command is the remote command to execute; argv array; not executed within a shell.
-	Command []string `json:"command" description:"the command to execute; argv array; not executed within a shell"`
+	Command []string `json:"command,csv" description:"the command to execute; argv array; not executed within a shell"`
```
Comment (bgrant0607) [inline pkg/api/v1/types.go]: csv doesn't belong to the json tag. There should be another tag, like urlparam or somesuch.

But don't do that yet, since I'm not convinced we should support the csv form.

File: pkg/apiserver/api_installer.go [hunk 3]
```diff
@@ -812,12 +812,14 @@ func typeToJSON(typeName string) string {
 		return "boolean"
 	case "uint8", "int", "int32", "int64", "uint32", "uint64":
 		return "integer"
-	case "byte":
-		return "string"
 	case "float64", "float32":
 		return "number"
+	case "byte":
```
Comment (bgrant0607) [inline pkg/apiserver/api_installer.go]: Why did you move this?

File: pkg/apiserver/api_installer.go [hunk 4]
```diff
@@ -812,12 +812,14 @@ func typeToJSON(typeName string) string {
...
+		return "string"
 	case "util.Time":
 		return "string"
+	case "[]string":
```
Comment (bgrant0607) [inline pkg/apiserver/api_installer.go]: This seems wrong to me. At minimum, there would need to be a comment explaining why this is doing this.

File: api/swagger-spec/v1.json (hunk 1 above)
Comment (bgrant0607) [inline api/swagger-spec/v1.json]: CSV doesn't seem very appropriate for commands.

File: pkg/apiserver/api_installer.go (hunk 3 above)
Comment (bgrant0607) [inline pkg/apiserver/api_installer.go]: Nevermind.

File: pkg/apiserver/api_installer.go (hunk 4 above)
Comment (bgrant0607) [inline pkg/apiserver/api_installer.go]: Ping.

File: api/swagger-spec/v1.json (hunk 1 above)
Comment (smarterclayton) [inline api/swagger-spec/v1.json]: As commented on the other issue CSV is incredibly dangerous - I would much prefer to say that swagger is broken, and we need to update to 2.0.  I don't think CSV makes the problem better.

File: pkg/api/v1/types.go (hunk 2 above)
Comment (nikhiljindal) [inline pkg/api/v1/types.go]: Moved the csv commit to a different PR: https://github.com/kubernetes/kubernetes/pull/13054

File: pkg/apiserver/api_installer.go (hunk 4 above)
Comment (nikhiljindal) [inline pkg/apiserver/api_installer.go]: I am trying to figure out a way to express this as 
```
"type": "array",
      "items": {
       "type": "string"
      },
```
https://github.com/emicklei/go-restful/issues/225

Will update the PR once I have it working

**Comment by nikhiljindal** [comment]
@bgrant0607 I have added a comment regarding why are we using "string" for a "[]string" query param.
Will fix that once we have a resolution for https://github.com/emicklei/go-restful/issues/225

I would like to submit this PR to fix the invalid swagger spec, being reported by users https://github.com/kubernetes/kubernetes/issues/13994

**Comment by bgrant0607** [comment]
LGTM

**Comment by thockin** [comment]
needs rebase

**Comment by nikhiljindal** [comment]
Rebased

**Comment by nikhiljindal** [comment]
Adding back the LGTM since it was just a sync and rebase
